### PR TITLE
provider/rundeck: Added multiple functionality.

### PR DESCRIPTION
**Pull request by BobVanB**
Added multiple functionality: schedule, descriptions, storagepath, scriptinterpreter, nodesSelectedByDefault, scriptinterpreter, nodefilterbydefault

Terraform plan is now clean, so only changes you made are seen.

UseCase: Changeing all our manual rundeck code to terraform, and missing some features in the process.

Still waiting for pull request on go-rundeck-api:
https://github.com/apparentlymart/go-rundeck-api/pull/9

### CODE REVIEW

**Comment by BobVanB** [comment]
Waiting for pulls on go-rundeck-api.

But this is fully functional code, i currently use in production.

```terraform
resource "rundeck_job" "my_job_name" {
  name = "My Job Name"
  description = "${file("${path.module}/scripts/description")}"
  project_name = "${rundeck_project.rundeck_project.name}"
  preserve_options_order = true

  node_filter_query = "${file("${path.module}/scripts/filter_query")}"
  max_thread_count = 10
  nodes_selected_by_default = true

  allow_concurrent_executions = false
  continue_on_error = false

  dispatch {
    continue_on_error = true
  }

  option {
    name = "Option"
    description = "${file("${path.module}/scripts/option_description")}"
    default_value = "status"
    value_choices = ["pulled","open","merged"]
    required = true
    require_predefined_choice = true
  }

  command {
    inline_script = "${file("${path.module}/scripts/inline_script")}"
    description = "Command Description"
  }
}
```

File: builtin/providers/rundeck/resource_job.go [hunk 1]
```diff
@@ -347,22 +415,80 @@ func jobFromResourceData(d *schema.ResourceData) (*rundeck.JobDetail, error) {
...
+		dispatchMap := dispatchConfigI[0].(map[string]interface{})
+		job.Dispatch.ContinueOnError = dispatchMap["continue_on_error"].(bool)
+	}
+
+	// Element: Job>Schedule
+	scheduleCron := d.Get("schedule_cron").(string)
+	if len(scheduleCron) > 0 {
+		scheduleCronArray := strings.Split(scheduleCron, " ")
+		if len(scheduleCronArray) != 7 {
+			return nil, fmt.Errorf("rundeck schedule_cron format is incorrect")
```
Comment (apparentlymart) [inline builtin/providers/rundeck/resource_job.go]: Could we use a `ValidateFunc` on the `schedule_cron` attribute for this? If we do the validation there then Terraform can detect it during the "plan" step in some cases (as long as it doesn't contain any interpolation expressions).

**Comment by apparentlymart** [comment]
Thanks for this, @BobVanB!

I merged the upstream client library change, and I left an initial comment here for now, but generally this looks good so far. Let me know when you're done here, and if you need any help with the acceptance tests or docs.

**Comment by BobVanB** [comment]
@apparentlymart I will pick this up monday, thnx for the review.

**Comment by BobVanB** [comment]
Do you mean this?
https://github.com/hashicorp/terraform/pull/8103

**Comment by BobVanB** [comment]
Added documentation and an advanced acceptance test.

As for the ValidateFunc, there are numerous checks done by rundeck. These errors are easy to understand when you do an apply plan. And are easy to find and modify.

In my opinion, i kept the number of logical checks to a minimum and just let rundeck do its job.

File: builtin/providers/rundeck/resource_job.go (hunk 1 above)
Comment (BobVanB) [inline builtin/providers/rundeck/resource_job.go]: As for the ValidateFunc, there are numerous checks done by rundeck. These errors are easy to understand when you do an apply plan. And are easy to find and modify.

In my opinion, i kept the number of logical checks to a minimum and just let rundeck do its job.

**Comment by BobVanB** [comment]
`

**Comment by BobVanB** [comment]
Schema has a MaxItems for List elements, will test en add this tomorrow.
This will validate the schema at plan setup.
Also added a deprication warning for exclude presedence.

**Comment by BobVanB** [comment]
@apparentlymart I think this one can be reviewed and is ready for merging.

**Comment by BobVanB** [comment]
I have another idea, but that has to be separate from this pull.
Write a wrapper around: https://github.com/asaskevich/govalidator
And put more validation in the schema.

**Comment by BobVanB** [comment]
I see that i dont have the latest go-rundeck-api version, updating it and then i will remove the wip

**Comment by BobVanB** [comment]
Currently adding errorhandling functionality.
If you have comments on changing keys, or other tips and stuff.
Just shout it out, there welcome and needed.
Else i am going to do what i think is easy. :D

**Comment by apparentlymart** [comment]
Hi @BobVanB!

I do intend to review this in more detail soon, but I just wanted to make a general note now: it seems like this patch is becoming quite a large set of changes, and the larger it gets the more complicated it will be to review. When possible it's preferable for each PR to add one specific feature so that it can be reviewed and merged in isolation from the rest.

I certainly don't want to discourage your momentum here, since there's some good stuff! Just want to make sure we can get this stuff merged efficiently, rather than going back and forth on many small details in a large patch.

**Comment by BobVanB** [comment]
Yeah I firgure that, was feeling the same.

The last one errorhandling changes alot of code.
I thing I should have put this seperatly.

Also still dont like that booleans are always present, there should be a third type for it "nil"
Omitempty an false can messup the plan output with unnessacary changes. But this will need an api change.

**Comment by BobVanB** [comment]
For now added yet another feature, ^^ have to wait for the api first :P 
notification>email/webhook

**Comment by BobVanB** [comment]
@apparentlymart After i checken out the boolean bug. I will cut this one Into smaller party.

**Comment by BobVanB** [comment]
Splitting this merge into multiple merges. closing this one.

**Pull request closed without merging**